routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,187 @@
+  AdjacentEdgesMap const & m_adjacentEdges;
+  TUnpackedPathSegments const & m_pathSegments;
+  double m_routeLength;
+};
+
+ftypes::HighwayClass GetHighwayClass(FeatureID const & featureId, Index const & index)
+{
+  ftypes::HighwayClass highWayClass = ftypes::HighwayClass::Undefined;
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  uint32_t const featureIndex = featureId.m_index;
+
+  FeatureType ft;
+  Index::FeaturesLoaderGuard loader(index, mwmId);
+  loader.GetFeatureByIndex(featureIndex, ft);
+  highWayClass = ftypes::GetHighwayClass(ft);
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Error, ());
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Undefined, ());
+  return highWayClass;
+}
+}  // namespace
+
+namespace routing
+{
+BicycleDirectionsEngine::BicycleDirectionsEngine(Index const & index) : m_index(index) {}
+
+void BicycleDirectionsEngine::Generate(IRoadGraph const & graph, vector<Junction> const & path,
+                                       Route::TTimes & times, Route::TTurns & turns,
+                                       vector<m2::PointD> & routeGeometry,
+                                       my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_NOT_EQUAL(pathSize, 0, ());
+
+  times.clear();
+  turns.clear();
+  routeGeometry.clear();
+  m_adjacentEdges.clear();
+  m_pathSegments.clear();
+
+  auto emptyPathWorkaround = [&]()
+  {
+    turns.emplace_back(pathSize - 1, turns::TurnDirection::ReachedYourDestination);
+    this->m_adjacentEdges[0] = AdjacentEdges(1);  // There's one ingoing edge to the finish.
+  };
+
+  if (pathSize <= 1)
+  {
+    ASSERT(false, (pathSize));
+    emptyPathWorkaround();
+    return;
+  }
+
+  CalculateTimes(graph, path, times);
+
+  IRoadGraph::TEdgeVector routeEdges;
+  if (!ReconstructPath(graph, path, routeEdges, cancellable))
+  {
+    LOG(LDEBUG, ("Couldn't reconstruct path"));
+    emptyPathWorkaround();
+    return;
+  }
+  if (routeEdges.empty())
+  {
+    ASSERT(false, ());
+    emptyPathWorkaround();
+    return;
+  }
+
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(0, AdjacentEdges(0)));
+  for (size_t i = 1; i < pathSize; ++i)
+  {
+    Junction const & prevJunction = path[i - 1];
+    Junction const & currJunction = path[i];
+    IRoadGraph::TEdgeVector outgoingEdges, ingoingEdges;
+    graph.GetOutgoingEdges(currJunction, outgoingEdges);
+    graph.GetIngoingEdges(currJunction, ingoingEdges);
+
+    AdjacentEdges adjacentEdges = AdjacentEdges(ingoingEdges.size());
+    adjacentEdges.m_outgoingTurns.reserve(outgoingEdges.size());
+    for (auto const & edge : outgoingEdges)
+    {
+      auto const & featureId = edge.GetFeatureId();
+      // Checking for if |edge| is a fake edge.
+      if (!featureId.m_mwmId.IsAlive())
+        continue;
+      double const angle = turns::PiMinusTwoVectorsAngle(
+          currJunction.GetPoint(), prevJunction.GetPoint(), edge.GetEndJunction().GetPoint());
```

Comment:
I did something else. Now this |angle| is not used at all for bicycle turn generation.